cpp/daal/src/algorithms/dtrees/forest/regression/df_regression_train_dense_default_impl.i

```diff
@@ -164,67 +171,80 @@ private:
+bool OrderedRespHelper<algorithmFPType, cpu>::init(const NumericTable * data, const NumericTable * resp, const IndexType * aSample,
+                                                   const NumericTable * weights)
 {
-    DAAL_CHECK_STATUS_VAR(super::init(data, resp, aSample));
+    DAAL_CHECK_STATUS_VAR(super::init(data, resp, aSample, weights));
     if (this->_indexedFeatures)
     {
         //init work buffer for the computation using indexed features
         const auto nDiffFeatMax = this->indexedFeatures().maxNumIndices();
         _idxFeatureBuf.reset(nDiffFeatMax);
-        return _idxFeatureBuf.get();
+        _weightsFeatureBuf.reset(nDiffFeatMax);
+        return _idxFeatureBuf.get() && _weightsFeatureBuf.get();
     }
     return true;
 }
 
 template <typename algorithmFPType, CpuType cpu>
-void OrderedRespHelper<algorithmFPType, cpu>::calcImpurity(const IndexType * aIdx, size_t n, ImpurityData & imp) const
+void OrderedRespHelper<algorithmFPType, cpu>::calcImpurity(const IndexType * aIdx, size_t n, ImpurityData & imp, algorithmFPType & totalWeights) const
 {
-    imp.var  = 0;
-    imp.mean = this->_aResponse[aIdx[0]].val;
+    imp.var      = 0;
+    imp.mean     = this->_aResponse[aIdx[0]].val;
+    totalWeights = this->_aWeights[aIdx[0]].val;
+    PRAGMA_VECTOR_ALWAYS
     for (size_t i = 1; i < n; ++i)
     {
-        algorithmFPType delta = this->_aResponse[aIdx[i]].val - imp.mean; //x[i] - mean
-        imp.mean += delta / algorithmFPType(i + 1);
-        imp.var += delta * (this->_aResponse[aIdx[i]].val - imp.mean);
+        const algorithmFPType weights = this->_aWeights[aIdx[i]].val;
+        const algorithmFPType delta   = this->_aResponse[aIdx[i]].val - imp.mean; //x[i] - mean
+        totalWeights += weights;
+        imp.mean += weights * delta / totalWeights;
+        imp.var += weights * delta * (this->_aResponse[aIdx[i]].val - imp.mean);
     }
-    imp.var /= algorithmFPType(n); //impurity is MSE
+    imp.var /= totalWeights; //impurity is MSE
 
 #ifdef DEBUG_CHECK_IMPURITY
-    TResponse mean1 = this->_aResponse[aIdx[0]].val / algorithmFPType(n);
-    for (size_t i = 1; i < n; ++i) mean1 += this->_aResponse[aIdx[i]].val / algorithmFPType(n);
-    algorithmFPType var1 = 0;
-    for (size_t i = 0; i < n; ++i) var1 += (this->_aResponse[aIdx[i]].val - mean1) * (this->_aResponse[aIdx[i]].val - mean1);
-    var1 /= algorithmFPType(n); //impurity is MSE
-    DAAL_ASSERT(fabs(mean1 - imp.mean) < 0.001);
-    DAAL_ASSERT(fabs(var1 - imp.var) < 0.001);
+    if (!this->_weights)
+    {
+        TResponse mean1 = this->_aResponse[aIdx[0]].val / algorithmFPType(n);
+        for (size_t i = 1; i < n; ++i) mean1 += this->_aResponse[aIdx[i]].val / algorithmFPType(n);
+        algorithmFPType var1 = 0;
+        for (size_t i = 0; i < n; ++i) var1 += (this->_aResponse[aIdx[i]].val - mean1) * (this->_aResponse[aIdx[i]].val - mean1);
+        var1 /= algorithmFPType(n); //impurity is MSE
```

Comment:
What about n==0 vs SDL?

Reply:
There is no problem in this current place because calcImpurity(const IndexType * aIdx, size_t n, ImpurityData & imp, algorithmFPType & totalWeights) method is called only with n>=1 value. This method is called in findBestSplitCategoricalFeature() on 518 row